lib/scala/logging-service-logback/src/main/java/org/enso/logging/service/logback/telemetry/TelemetryAppender.java

```diff
@@ -0,0 +1,270 @@
+          .build();
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Transforms the given log events into JSON payloads.
+   *
+   * @return null if none of the log events could be transformed into a payload.
+   */
+  private String buildPayload(List<ILoggingEvent> logEvents) {
+    var payload = new ObjectNode(JsonNodeFactory.instance);
+    var logs = new ArrayNode(JsonNodeFactory.instance);
+    for (var logEvent : logEvents) {
+      var logMessage =
+          new LogMessage(
+              logEvent.getLoggerName(), logEvent.getMessage(), logEvent.getArgumentArray());
+      var payloadForLogEvent = LogFormatter.transform(logMessage);
+      if (payloadForLogEvent != null) {
+        logs.add(payloadForLogEvent);
+      }
+    }
+    if (logs.size() != logEvents.size()) {
+      LOGGER.warn("Failed to build payload for some log events");
+    }
+    if (logs.isEmpty()) {
+      return null;
+    } else {
+      payload.set("logs", logs);
+      return payload.toString();
+    }
+  }
+
+  private void sendLogRequest(HttpRequest request, int retryCount) throws RequestFailureException {
+    assert request != null;
+    try {
+      try {
+        if (httpClient == null) {
+          httpClient = HttpClient.newBuilder().followRedirects(HttpClient.Redirect.ALWAYS).build();
+        }
+        HttpResponse<String> response =
+            httpClient.send(request, HttpResponse.BodyHandlers.ofString());
+        if (response.statusCode() < 200 || response.statusCode() >= 300) {
+          throw new RequestFailureException(
+              "Unexpected status code: " + response.statusCode() + " " + response.body(), null);
+        }
+      } catch (IOException | InterruptedException e) {
+        // Promote a checked exception to a runtime exception to simplify the code.
+        var errorMessage = e.getMessage() != null ? e.getMessage() : e.toString();
+        throw new RequestFailureException("Failed to send log messages: " + errorMessage, e);
+      }
+    } catch (RequestFailureException e) {
+      if (retryCount < 0) {
+        LOGGER.debug("Failed to send log messages after retrying", e);
+        throw e;
+      } else {
+        LOGGER.debug("Exception when sending log messages. Retrying...", e);
+        sendLogRequest(request, retryCount - 1);
+      }
+    }
+  }
+
+  private static final class RequestFailureException extends Exception {
+    public RequestFailureException(String message, Throwable cause) {
+      super(message, cause);
+    }
+  }
+
+  /** The credentials file is created by the IDE once user logs in. We are just reading it. */
+  private record Credentials(
+      @JsonProperty("client_id") String clientId,
+      @JsonProperty("access_token") String accessToken,
+      @JsonProperty("refresh_token") String refreshToken,
+      @JsonProperty("refresh_url") String refreshUrl,
+      @JsonProperty("expire_at") String expireAt) {}
```

Comment:
I see the logging engine is lacking all the refresh token logic that Enso [`Authentication`](https://github.com/enso-org/enso/blob/fe8e1de5ab5e73c6a6705f4a813b21187054e96f/distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Internal/Authentication.enso#L4) is handling.

That means that if a project stays open for a long time, Libraries Cloud integration will keep working (as it will refresh the token itself) but the telemetry logs will cease to be sent after some interval (once the access token expires). 

I guess we can alleviate that by extending the default lifetime of the access token, but please do keep that in mind.

Reply:
If we created a common 'base cloud support' lib for engine and libs that could contain:
- getting API url from env vars
- locating, parsing and refreshing credentials
Then we could be safe from such situations.

Reply:
Tracked as a follow-up in https://github.com/enso-org/enso/issues/12564